### PR TITLE
feat(core-flows,medusa,pricing,types,utils): added price list workflows + endpoints

### PR DESCRIPTION
**Pull request by riqwan**
Apologies for the giant PR in advance, but most of these are removing of files and migrations from old workflows and routes to new.

What:

- Adds CRUD endpoints for price lists
- Migrate workflows from old sdk to new
- Added missing updatePriceListPrices method to pricing module

### CODE REVIEW

File: packages/core-flows/src/price-list/steps/delete-price-lists.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    if (!idsToRestore?.length) {
+      return
+    }
+
+    const pricingModule = container.resolve<IPricingModuleService>(
+      ModuleRegistrationName.PRICING
+    )
+
+    // TODO: Implement restore price lists
+    // await pricingModule.restorePriceLists(idsToRestore)
```
Comment (olivermrbl) [inline packages/core-flows/src/price-list/steps/delete-price-lists.ts:28]: question: Is there anything to implement? If you add it to the interface, you should be able to use it here, and the abstract module service will take care of the rest.

File: packages/types/src/pricing/common/price-list.ts [hunk 2]
```diff
@@ -416,7 +430,23 @@ export interface AddPriceListPricesDTO {
...
+/**
+ * @interface
+ *
+ * The prices to be added to a price list.
+ */
+export interface UpdatePriceListPricesDTO {
+  /**
+   * The ID of the price list to add prices to.
+   */
+  priceListId: string
```
Comment (olivermrbl) [inline packages/types/src/pricing/common/price-list.ts]: todo: snake case
```suggestion
  price_list_id: string
```

File: packages/medusa/src/api-v2/admin/price-lists/route.ts [hunk 3]
```diff
@@ -27,3 +30,32 @@ export const GET = async (
...
+  const { result, errors } = await workflow.run({
+    input: { priceListsData },
+    throwOnError: false,
+  })
+
+  if (Array.isArray(errors) && errors[0]) {
+    throw errors[0].error
+  }
+
+  const [[priceList]] = await listPriceLists({
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/route.ts:52]: suggestion(non-blocking): Use `remoteQuery`

File: packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts [hunk 4]
```diff
@@ -29,3 +31,77 @@ export const GET = async (
...
+  const id = req.params.id
+  const workflow = deletePriceListsWorkflow(req.scope)
+
+  const [_, count] = await listPriceLists({
+    container: req.scope,
+    fields: ["id"],
+    variables: { filters: { id }, skip: 0, take: 1 },
+  })
+
+  if (count === 0) {
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts]: todo: Should be idempotent, just return:
```
{
    id,
    object: "price_list",
    deleted: true,
  }
```

...even though the ID didn't exist.

File: packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts [hunk 5]
```diff
@@ -29,3 +31,77 @@ export const GET = async (
...
+  const { errors } = await workflow.run({
+    input: { priceListsData: [{ id, ...req.validatedBody }] },
+    throwOnError: false,
+  })
+
+  if (Array.isArray(errors) && errors[0]) {
+    throw errors[0].error
+  }
+
+  const [[priceList]] = await listPriceLists({
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts:59]: suggestion(non-blocking): Use `remoteQuery`

File: packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts [hunk 6]
```diff
@@ -29,3 +31,77 @@ export const GET = async (
...
+    fields: ["id"],
+    variables: { filters: { id }, skip: 0, take: 1 },
+  })
+
+  if (count === 0) {
+    throw new MedusaError(
+      MedusaError.Types.NOT_FOUND,
+      `Price list with id: ${id} was not found`
+    )
+  }
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/[id]/route.ts]: thought(non-blocking): Would it make sense to move these to the workflow level?

What happens if the price list does not exist in the workflow? Do we handle that gracefully?

File: packages/core-flows/src/price-list/workflows/update-price-lists.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { PriceListDTO, UpdatePriceListWorkflowInputDTO } from "@medusajs/types"
+import { WorkflowData, createWorkflow } from "@medusajs/workflows-sdk"
+import { updatePriceListsStep } from "../steps"
+
+type WorkflowInput = { priceListsData: UpdatePriceListWorkflowInputDTO[] }
```
Comment (olivermrbl) [inline packages/core-flows/src/price-list/workflows/update-price-lists.ts]: todo: snake case
```suggestion
type WorkflowInput = { price_list_data: UpdatePriceListWorkflowInputDTO[] }
```

File: packages/core-flows/src/price-list/workflows/create-price-lists.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { CreatePriceListWorkflowInputDTO, PriceListDTO } from "@medusajs/types"
+import { WorkflowData, createWorkflow } from "@medusajs/workflows-sdk"
+import { createPriceListsStep } from "../steps"
+
+type WorkflowInput = { priceListsData: CreatePriceListWorkflowInputDTO[] }
```
Comment (olivermrbl) [inline packages/core-flows/src/price-list/workflows/create-price-lists.ts]: todo: snake case

File: packages/core-flows/src/price-list/steps/delete-price-lists.ts (hunk 1 above)
Comment (riqwan) [reply packages/core-flows/src/price-list/steps/delete-price-lists.ts:28]: Yes, unfortunately the models within aren't soft deletable-ready. Some of them don't even have the deleted_at attribute

File: packages/medusa/src/api-v2/admin/price-lists/route.ts (hunk 3 above)
Comment (riqwan) [reply packages/medusa/src/api-v2/admin/price-lists/route.ts:52]: This function is already calling remote query inside

Comment (olivermrbl) [reply packages/medusa/src/api-v2/admin/price-lists/route.ts:52]: Ah my bad. Discard this and the other comments then.

**Comment by riqwan** [comment]
@olivermrbl addressed all your comments, shall we merge this in if it looks good?

File: integration-tests/modules/__tests__/price-lists/admin/price-lists.spec.ts [hunk 9]
```diff
@@ -258,6 +258,349 @@ medusaIntegrationTestRunner({
...
+                  rules: {
+                    region_id: region.id,
+                  },
+                }),
+              ],
+            })
+          )
+        })
+
+        it("should should throw error when required attributes are not provided", async () => {
```
Comment (olivermrbl) [inline integration-tests/modules/__tests__/price-lists/admin/price-lists.spec.ts]: nit:
```suggestion
        it("should throw error when required attributes are not provided", async () => {
```

File: integration-tests/modules/__tests__/price-lists/admin/price-lists.spec.ts [hunk 10]
```diff
@@ -258,6 +258,349 @@ medusaIntegrationTestRunner({
...
+            data,
+            adminHeaders
+          )
+
+          expect(response.status).toEqual(200)
+          expect(response.data.price_list.prices.length).toEqual(2)
+          expect(response.data.price_list).toEqual(
+            expect.objectContaining({
+              id: expect.any(String),
+              prices: expect.arrayContaining([
```
Comment (olivermrbl) [inline integration-tests/modules/__tests__/price-lists/admin/price-lists.spec.ts:550]: suggestion: Assert on price ID

File: packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+import {
+  adminPriceListRemoteQueryFields,
+  defaultAdminPriceListFields,
+} from "../../query-config"
+import {
+  AdminDeletePriceListsPriceListPricesReq,
+  AdminPostPriceListsPriceListPricesReq,
+} from "../../validators"
+
+export const POST = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts:19]: suggestion: Should we move this to `POST /admin/price-lists/:id/prices/update`?

File: packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    container: req.scope,
+    remoteQueryFields: adminPriceListRemoteQueryFields,
+    apiFields: defaultAdminPriceListFields,
+    variables: { filters: { id }, skip: 0, take: 1 },
+  })
+
+  res.status(200).json({ price_list: priceList })
+}
+
+export const DELETE = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts:47]: suggestion: Should we move this to `POST /admin/price-lists/:id/prices/remove`?

File: packages/medusa/src/api-v2/admin/price-lists/middlewares.ts [hunk 13]
```diff
@@ -33,4 +37,24 @@ export const adminPriceListsRoutesMiddlewares: MiddlewareRoute[] = [
...
+  {
+    method: ["POST"],
+    matcher: "/admin/price-lists/:id/prices",
+    middlewares: [transformBody(AdminPostPriceListsPriceListPricesReq)],
+  },
+  {
+    method: ["DELETE"],
+    matcher: "/admin/price-lists/:id/prices",
+    middlewares: [transformBody(AdminDeletePriceListsPriceListPricesReq)],
+  },
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/price-lists/middlewares.ts:59]: See other comment on endpoint paths

**Review by olivermrbl — APPROVED** [review]
LGTM, just a couple of things we need to align on before we merge

File: packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts (hunk 12 above)
Comment (riqwan) [reply packages/medusa/src/api-v2/admin/price-lists/[id]/prices/route.ts:47]: Yes, it's already part of the work I'm doing next along with update prices.

**Comment by riqwan** [comment]
@olivermrbl will bring over the route changes to the next PR. 👍🏻 will stick to smaller PRs from here onward 😄

**Comment by olivermrbl** [comment]
sg!

**Pull request merged**